### ISSUE TITLE
add per-file configuration support in configuration files.

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#511]
In GitLab by @mjpieters on Jul 4, 2016, 04:19

Currently the only way to disable specific warnings in a specific file is to add `# noqa` entries per offending line. Please add support for suppressing specific errors for specific files, where *all* instances of a specific error within that file are ignored.

For example, in polyglot projects (supporting Py2 + Py3) a pycompat file is commonplace that does a lot of imports that are not used *in that module*. Having to mark this file as completely exempted from linting leaves the file open to other linting errors, and marking each import with `# noqa` gets tedious fast.

I propose that a header with `# flake8 noqa: error_code, error_code, etc.` acts the same as if `# noqa: error_code, error_code, etc.` has been added to every code line in the file, and / or that a `[relative/file/path]` or `[**/globbed/filepath.*]` section in a configuration file allows you to specify additional configuration for a given file or set of files.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jul 5, 2016, 13:02

So this is something that has been requested before. Currently flake8-putty fills this needs.

This is certainly not something that will make it into 3.0.0 but I may change my mind and include it in a later 3.x release. I'm pretty sure it would be a config-file only way of specifying these per-file ignores.

To be clear, I'm very much against this. It undermines the very idea of enforcing a Style Guide on a project (which is Flake8's fundamental purpose). Having per-file sets of rules, defeats the idea of a per-project style guide.

**Comment by asottile** [comment]
In GitLab by @The-Compiler on Jul 5, 2016, 13:23

Isn't this a duplicate of #143?

Apart from the monorepo-thing @ahal mentioned there, I think another valid usecase is wanting to ignore some error codes for tests.

For example, I ignore `D100` and `D101` in test files (via `flake8-putty` currently). I want to enforce docstrings for my code, but for some simple tests something like:

```python
def test_the_function():
    """Test if the_function returns the right thing."""
    assert the_function() == ...
    # there can even be some more code here, I think the point still stands for some tests
```

the docstring just adds noise and no value, which is why I don't enforce the same rule for tests.

**Comment by asottile** [comment]
In GitLab by @reedstrm on Sep 23, 2016, 08:00

Chiming in here with my support for the use-case. Tests are the classic - sometimes you have to do things in tests that are not good python-coding style. Our case is loooong lines of expect output, to match the output from underlying tools.  These are currently cut-&-paste into the tests - any more proper way (concatenate consecutive lines of quoted strings, read an external file, etc.) would probably mean the developers are less likely to write that test. :-(

**Comment by asottile** [comment]
In GitLab by @saifelse on May 14, 2017, 21:53

In my organization's codebase, we'd like to adopt the style to allow unused imports in `__init__.py`. It's unclear based on [this pyflake's thread](https://bugs.launchpad.net/pyflakes/+bug/1178905) if there will be support for configuring the lint rule at a more granular level that would support our need.

Despite this, the above mentioned flake8 feature would allow us to workaround the issue... e.g. if we were able to specify in our `setup.cfg` something like:

```
[flake8.overrides.myrule1]
filename-re = ^(.*/)?__init__.py$
ignore = F401  # ideally there is a nice syntax to make this in addition to the normal set. 
```

**Comment by asottile** [comment]
In GitLab by @luzfcb on Aug 4, 2017, 07:33

this is the exact feature to solve this problem: https://gitlab.com/pycqa/flake8/issues/146
Is there any plan to include this in the near future?

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 5, 2017, 05:30

@luzfcb that's what the "feature:accepted" label means. If the description of that label doesn't make that clear, please let us know how we can improve it.

**Comment by asottile** [comment]
In GitLab by @blueyed on May 16, 2018, 11:11

For the record: flake8-putty does not work with flake8>3, but there is https://github.com/snoack/flake8-per-file-ignores.

**Comment by asottile** [comment]
In GitLab by @matthijskooijman on May 19, 2018, 11:56

Note that the in-file (`#flake8: noqa: E123`) approach suggested as one alternative in the original post is also the subject of #399 (so it probably makes sense to keep this issue about the configuration file support only).

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 24, 2018, 08:49

assigned to @sigmavirus24

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 25, 2018, 07:13

mentioned in merge request !259

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 27, 2018, 10:01

People interested in this should test out !259

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 29, 2018, 08:37

closed via commit f2776107db605eb67362db00bc5e9ef2afa9f39e
